### PR TITLE
Added code and WDL to complete ModelSegments CNV pipeline.

### LINKED ISSUE #2858: GATK CNV on WGS.
@LeeTL1220 commented on [Mon May 23 2016](https://github.com/broadinstitute/gatk-protected/issues/524)

- [ ] File issues for evaluations (and make sure each evaluation is described, including ground truth).  Make sure that there is a milestone and assignee for each.

Need to show some measure of performance for the WGS evaluation (GATK CNV only.  Not ACNV)

### LINKED ISSUE #3661: Update CNV WDLs to use PreprocessIntervals and CollectReadCounts.
@MartonKN, this will be a good introduction to writing WDL.  @asmirnov239 and I can give you some pointers to start on this after he finishes #3660.

### LINKED ISSUE #3825: Make and evaluate simple changes to ReCapSegCaller.
Some preliminary evaluation of the new ModelSegments pipeline on CRSP samples has revealed some weaknesses of the ReCapSeg caller (which is simply ported from the old pipeline) to me.

I think there are a lot of confusing things going on:

1) For determining copy-neutral segments, all segments with log2 mean below some threshold are used (rather than absolute log2).  There is a comment that this is done to "mimic the python code" but I have no idea why this would be sensible, since it includes all deletions.
2) There is some confusion arising from inconsistent use of z-score and T-statistic.  Standard deviation, rather than standard error, is used for calling; i.e., a "called segment" is one that has a mean log2 copy ratio that has a z-score above some threshold with respect to the standard deviation of the log2 copy ratios of intervals that fall within copy-neutral segments (note also that these intervals have already been filtered by z-score to remove outliers).  That is, any segment with a mean that falls sufficiently within the fuzziness of the caterpillar is not called.
3) However, even calling using standard error is probably not what we want.  This would simply be asking the question: given a population of copy-neutral intervals with a mean and standard deviation, does any non-copy-neutral segment contain intervals with a mean significantly different than the population?  We've already answered this question during segmentation!

I think what we want to do instead is ask questions about the population of segment-level copy-ratio estimates, weighted by length.

### PR DESCRIPTION
**Pull request by samuelklee**
Some notes on individual commits:

Updated CallCopyRatioSegments and PreprocessIntervals; reorganized copynumber packages.
-For motivation of changes in CallCopyRatioSegments, see #3825.
-I added the ability to turn off binning in PreprocessIntervals by specifying bin_length = 0.
-I removed the separation between coverage and allelic packages to make the package structure a bit simpler.
-@MartonKN should review, since he wrote PreprocessIntervals and is updating the caller.

Added segmentation classes and tests for ModelSegments CNV pipeline.
-I added implementations of copy-ratio, allele-fraction, and "multidimensional" (joint) segmentation.  All implementations are pretty boilerplate; they simply partition by contig and then call out to KernelSegmenter.  Note that there is some logic in multidimensional segmentation that only uses the first het in each copy-ratio interval and if any are available, and imputes the alt-allele fraction to 0.5 if not.
-Makes sense for @mbabadi to review this, since he reviewed the KernelSegmenter PR.

Added modeling classes and tests for ModelSegments CNV pipeline.
-Most of this code is copied from the old MCMC code.  However, I've done some overall code cleanup and refactoring, especially to remove some overextraction of methods in the allele-fraction likelihoods (see #2860).  I also added downsampling and scaling of likelihoods to cut down on runtime.  Tests have been simplified and rewritten to use simulated data.
-@LeeTL1220 do you think you could take a look?

Added ModelSegments CLI.
-Mostly control flow to handle optional inputs and validation, but there is some ugly and not well documented code that essentially does the GetHetCoverage step.  We'll refactor later, I filed #3915.
-@asmirnov239 can review.  This is lower priority than the gCNV VCF writing.

Deleted gCNV WDL and Cromwell tests.
-Trivial to review.

Added WDL and Cromwell tests for ModelSegments CNV pipeline.
-This includes the cost optimizations from @meganshand and @jsotobroad (sorry guys, I wasn't sure how to track your contributions while fixing up commits!)  I also added tests for both GC/no-GC pair workflows.
-@MartonKN should review to gain familiarity with the WDL.  Note that this WDL has already been through many revisions from @meganshand, @jsotobroad, and @LeeTL1220, so hopefully there shouldn't be too much for you to find serious fault with.

Note that I punted on adding MultidimensionalKernelSegmenterUnitTest and ModelSegmentsIntegrationTest.  Filed #3916.

Closes #2858.  (FINALLY!)
Closes #3825.
Closes #3661.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+
+    @Override
+    protected Object doWork() {
+        validateArguments();
+
+        //perform one-dimensional or multidimensional segmentation as appropriate and write to file
+        //(for use by CallCopyRatioSegments, if copy ratios are available)
+        final MultidimensionalSegmentCollection multidimensionalSegments;
+        if (inputDenoisedCopyRatiosFile != null && inputAllelicCountsFile == null) {
+            readDenoisedCopyRatios();
```
Comment (asmirnov239) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: I do prefer more functional code and I feel like this and other methods calls below (such as `readDenoisedCopyRatios` and `readAndFilterAllelicCounts`) should not work through side effects

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+            final AllelicCountCollection unfilteredNormalAllelicCounts = new AllelicCountCollection(inputNormalAllelicCountsFile);
+            if (!unfilteredNormalAllelicCounts.getIntervals().equals(unfilteredAllelicCounts.getIntervals())) {
+                throw new UserException.BadInput("Allelic-count sites in case sample and matched normal do not match. " +
+                        "Run CollectAllelicCounts using the same interval list of sites for both samples.");
+            }
+            final SampleMetadata normalSampleMetadata = unfilteredNormalAllelicCounts.getSampleMetadata();
+
+            //filter on total count in matched normal
+            logger.info(String.format("Filtering allelic counts in matched normal with total count less than %d...", minTotalAlleleCount));
+            final AllelicCountCollection filteredNormalAllelicCounts = new AllelicCountCollection(
```
Comment (asmirnov239) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: You could extract this and other similar filtering code to a new function.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+
+import java.io.File;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: Yup.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java (hunk 1 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: I agree, the control flow could probably be improved here and in the gCNV CLIs.  Filed #3951.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+            readAndFilterAllelicCounts();
+            multidimensionalSegments = new MultidimensionalKernelSegmenter(denoisedCopyRatios, hetAllelicCounts)
+                    .findSegmentation(maxNumSegmentsPerChromosome,
+                            kernelVarianceCopyRatio, kernelVarianceAlleleFraction, kernelScalingAlleleFraction, kernelApproximationDimension,
+                            ImmutableSet.copyOf(windowSizes).asList(),
+                            numChangepointsPenaltyFactor, numChangepointsPenaltyFactor);
+        }
+
+        logger.info("Modeling available denoised copy ratios and heterozygous allelic counts...");
+        //initial MCMC model fitting performed by MultidimensionalModeller constructor
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: Can you do this in non-Javadoc comments?

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+                        .collect(Collectors.toList()));
+        writeSegments(copyRatioSegmentsFinal, COPY_RATIO_SEGMENTS_FOR_CALLER_FILE);
+
+        logger.info("SUCCESS: ModelSegments run complete.");
+
+        return "SUCCESS";
+    }
+
+    private void validateArguments() {
+        Utils.nonNull(outputPrefix);
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: Good point, will also double check that this is done in corresponding tools.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java (hunk 2 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: Yup, already filed #3915.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java]: It is, and done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java [hunk 6]
```diff
@@ -61,15 +59,46 @@
...
+            fullName = NEUTRAL_SEGMENT_COPY_RATIO_THRESHOLD_LONG_NAME,
+            shortName = NEUTRAL_SEGMENT_COPY_RATIO_THRESHOLD_SHORT_NAME,
+            optional = true
+    )
+    private double neutralSegmentCopyRatioThreshold = 0.1;
+
+    @Argument(
+            doc = "Threshold on z-score of non-log2 copy ratio used for determining outlier copy-neutral segments.  " +
+                    "If non-log2 copy ratio z-score is above this threshold for a copy-neutral segment, " +
+                    "it is considered an outlier and not used in the calculation of the length-weighted standard deviation " +
```
Comment (MartonKN) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java]: I think it should say 'length-weighted mean and standard deviation used for calling the segment.'

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java [hunk 7]
```diff
@@ -61,15 +59,46 @@
...
+            fullName = OUTLIER_NEUTRAL_SEGMENT_COPY_RATIO_Z_SCORE_THRESHOLD_LONG_NAME,
+            shortName = OUTLIER_NEUTRAL_SEGMENT_COPY_RATIO_Z_SCORE_THRESHOLD_SHORT_NAME,
+            optional = true,
+            minValue = 0.
+    )
+    private double outlierNeutralSegmentCopyRatioZScoreThreshold = 2.;
+
+    @Argument(
+            doc = "Threshold on z-score of non-log2 copy ratio used for calling segments.  " +
+                    "If non-log2 copy ratio z-score is above this threshold for a copy-neutral segment, " +
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java]: I think I accidentally copied and pasted a few lines from above here.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java [hunk 8]
```diff
@@ -66,11 +67,11 @@
     public static final String PADDING_SHORT_NAME = "P";
 
     @Argument(
-            doc = "Length (in bp) of the bins.",
+            doc = "Length (in bp) of the bins.  If zero, no binning will be performed.",
```
Comment (MartonKN) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java]: Double space in front of 'If'

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java [hunk 9]
```diff
@@ -120,6 +121,9 @@ private static IntervalList padAndMergeIntervals(final List<SimpleInterval> inpu
     }
 
     private static IntervalList generateBins(final IntervalList preparedIntervalList, final int binLength, final SAMSequenceDictionary sequenceDictionary) {
+        if (binLength == 0) {
+            return IntervalList.copyOf(preparedIntervalList);
+        }
         final IntervalList bins = new IntervalList(sequenceDictionary);
         for (final Interval interval : preparedIntervalList) {
             for (int binStart = interval.getStart(); binStart <= interval.getEnd(); binStart += binLength) {
```
Comment (MartonKN) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java]: The changes look good!

**Comment by MartonKN** [comment]
Done with my code review @samuelklee!

**Comment by samuelklee** [comment]
@LeeTL1220, @mbabadi is out on vacation.  Would you mind reviewing the commit I assigned to him?  Thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/segmentation/MultidimensionalKernelSegmenter.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                        allelicCountOverlapDetector.getOverlaps(cr).stream()
+                                .min(SampleLocatableCollection.LEXICOGRAPHICAL_ORDER_COMPARATOR::compare)
+                                .orElse(BALANCED_ALLELIC_COUNT).getAlternateAlleleFraction()))
+                .collect(Collectors.groupingBy(
+                        MultidimensionalPoint::getContig,
+                        LinkedHashMap::new,
+                        Collectors.toList()));
+    }
+
+    public MultidimensionalSegmentCollection findSegmentation(final int maxNumChangepointsPerChromosome,
```
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/segmentation/MultidimensionalKernelSegmenter.java]: Missing javadoc

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java (hunk 7 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java]: Done.

File: .travis.yml [hunk 11]
```diff
@@ -101,7 +102,7 @@ install:
   else
     ./gradlew assemble;
     ./gradlew installDist;
-    if [[ $RUN_CNV_SOMATIC_LEGACY_WDL == true || $RUN_M2_WDL == true || $RUN_CNV_GERMLINE_WDL == true ]]; then
+    if [[ $RUN_CNV_SOMATIC_LEGACY_WDL == true || $RUN_CNV_SOMATIC_WDL == true || $RUN_CNV_GERMLINE_WDL == true || $RUN_M2_WDL == true ]]; then
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply .travis.yml]: This is cleaned up in a later PR.

File: .travis.yml [hunk 12]
```diff
@@ -116,12 +117,12 @@ script:
     echo "Not running any tests for nightly builds";
   elif [[ $TRAVIS_SECURE_ENV_VARS == false && $TEST_TYPE == cloud ]]; then
     echo "Can't run cloud tests without keys so don't run tests";
+  elif [[ $RUN_CNV_SOMATIC_WDL == true ]]; then
+    echo "Running CNV somatic workflows";
+    bash scripts/cnv_cromwell_tests/somatic/run_cnv_somatic_workflows.sh;
   elif [[ $RUN_CNV_SOMATIC_LEGACY_WDL == true ]]; then
     echo "Running legacy CNV somatic workflows";
     bash scripts/cnv_cromwell_tests/somatic_legacy/run_cnv_somatic_workflows.sh;
-  elif [[ $RUN_CNV_GERMLINE_WDL == true ]]; then
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply .travis.yml]: Also cleaned up in the germline PRs.  The new workflows are run there.

[the remark this answers is not in the document]
Comment (samuelklee) [reply .travis.yml]: FYI, gCNV Spark also got blown away.

File: scripts/cnv_wdl/cnv_common_tasks.wdl [hunk 13]
```diff
@@ -240,20 +93,23 @@ task CollectCounts {
     String base_filename = basename(bam, ".bam")
     String counts_filename = if !defined(output_format) then "${base_filename}.counts.hdf5" else "${base_filename}.counts.tsv"
 
-    command {
-        java -Xmx${default="8" mem}g -jar ${gatk_jar} CollectFragmentCounts \
+    command <<<
+        set -e
+        GATK_JAR=${default="/root/gatk.jar" gatk4_jar_override}
+
+        java -Xmx${default="8" mem}g -jar $GATK_JAR CollectFragmentCounts \
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply scripts/cnv_wdl/cnv_common_tasks.wdl]: I don't think he made the changes to all tasks.  Perhaps just the most memory intensive ones.  Made a note in #3940 to clean these up.

File: scripts/cnv_wdl/somatic/README.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+To get started, copy the relevant ``*.template.json`` for the workflow you wish to run and adjust parameters accordingly.  
+
+*Please note that there are optional workflow-level and task-level parameters that do not appear in the template files.  These are set to reasonable values by default, but can also be adjusted if desired.*
+
+#### Required parameters in the somatic panel workflow
+
+The reference used must be the same between PoN and case samples.
+
+- ``CNVSomaticPanelWorkflow.gatk_docker`` -- GATK Docker image (e.g., ``broadinstitute/gatk:x.beta.x``).
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply scripts/cnv_wdl/somatic/README.md]: Good catch.  Made a note to change this in the gCNV WDL readme too.

File: scripts/cnv_wdl/somatic/cnv_somatic_pair_workflow_template.json [hunk 15]
```diff
@@ -0,0 +1,13 @@
+{
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply scripts/cnv_wdl/somatic/cnv_somatic_pair_workflow_template.json]: These are just placeholders in the M2 JSON templates ("$__intervals__").  See the "Please note that python [templates](https://docs.python.org/2/library/string.html#template-strings)..." in the M2 WDL readme.  I think this was copied over from the legacy CNV readme, from which it was subsequently deleted.

This is essentially the output of wdltool inputs, and I think that having the parameter type is actually more useful than the python template strings.  But in my opinion, these templates are only slightly useful to begin with and they're one more thing that we have to keep up to date.  I'm actually fine with removing them altogether.  @davidbenjamin

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/segmentation/MultidimensionalKernelSegmenter.java (hunk 10 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/segmentation/MultidimensionalKernelSegmenter.java]: Good catch, added some here and to the other KernelSegmenters.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java]: Not sure what our policy is on double spaces between sentences.  @sooheelee @vdauwera?

File: scripts/cnv_wdl/somatic/cnv_somatic_pair_workflow_template.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply scripts/cnv_wdl/somatic/cnv_somatic_pair_workflow_template.json]: > But in my opinion, these templates are only slightly useful to begin with and they're one more thing that we have to keep up to date. I'm actually fine with removing them altogether.

I don't think they're useful, either.

[the remark this answers is not in the document]
Comment (samuelklee) [reply scripts/cnv_wdl/somatic/cnv_somatic_pair_workflow_template.json]: OK, will delete them.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java [hunk 16]
```diff
@@ -61,15 +59,46 @@
...
+                    "it is considered an outlier and not used in the calculation of the length-weighted standard deviation " +
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CallCopyRatioSegments.java]: Good catch, done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java]: Leaving them for now, then.

**Comment by samuelklee** [comment]
Thanks all, addressed comments and rebased.  Will merge after tests pass unless there are any objections.

**Comment by samuelklee** [comment]
Oops, pretty sure that I screwed up the rebase and that tests will fail.  Will fix later tonight.

**Pull request merged**